File: jacsgo/cross_language_provenance_test.go
// Cross-language provenance tests (Task 13, PRD §5.1 / §5.2).
//
// Verifies that text + image fixtures signed by Rust under
// jacs/tests/fixtures/provenance/ are accepted by the Go binding, and that
// a Go-signed file round-trips through the Rust `jacs verify-text` CLI.
//
// Fixtures are committed; regenerate with:
//
//   UPDATE_PROVENANCE_FIXTURES=1 cargo test -p jacs --test \
//     provenance_cross_language_tests -- --ignored regenerate_provenance_fixtures
//
// The whole file is skipped when fixtures are absent.

package jacs

import (
	"encoding/json"
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"gopkg.in/yaml.v3"
)

// fixturesDir resolves the absolute path to jacs/tests/fixtures/provenance/
// from the test package directory using runtime.Caller (more robust than the
// cwd, which varies across IDEs/CI).
func provenanceFixturesDir(t *testing.T) string {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	// jacsgo/cross_language_provenance_test.go → JACS/jacsgo → JACS → jacs/tests/fixtures/provenance
	return filepath.Clean(filepath.Join(filepath.Dir(thisFile), "..", "jacs", "tests", "fixtures", "provenance"))
}

func provenanceKeysDir(t *testing.T) string {
	return filepath.Join(provenanceFixturesDir(t), "keys")
}

type provenanceMetadata struct {
	Schema       string `json:"schema"`
	GeneratedBy  string `json:"generated_by"`
	JacsVersion  string `json:"jacs_version"`
	AgentEd25519 struct {
		AgentID            string `json:"agent_id"`
		Algorithm          string `json:"algorithm"`
		PublicKeyFilename  string `json:"public_key_filename"`
	} `json:"agent_ed25519"`
	AgentPq2025 struct {
		AgentID            string `json:"agent_id"`
		Algorithm          string `json:"algorithm"`
		PublicKeyFilename  string `json:"public_key_filename"`
	} `json:"agent_pq2025"`
}

func loadProvenanceMetadata(t *testing.T) provenanceMetadata {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(provenanceFixturesDir(t), "metadata.json"))
	if err != nil {
		t.Fatalf("read metadata.json: %v", err)
	}
	var m provenanceMetadata
	if err := json.Unmarshal(data, &m); err != nil {
		t.Fatalf("parse metadata.json: %v", err)
	}
	return m
}

// skipIfFixturesMissing skips the test if the committed fixture set is not
// materialised on disk. Other bindings ship a comparable skip; CI runs the
// regenerate step first when needed.
func skipIfFixturesMissing(t *testing.T) {
	t.Helper()
	meta := filepath.Join(provenanceFixturesDir(t), "metadata.json")
	keys := provenanceKeysDir(t)
	if _, err := os.Stat(meta); errors.Is(err, os.ErrNotExist) {
		t.Skip("provenance fixtures not generated; run UPDATE_PROVENANCE_FIXTURES=1 cargo test ...")
	}
	if _, err := os.Stat(keys); errors.Is(err, os.ErrNotExist) {
		t.Skip("provenance keys directory missing")
	}
}

// ---------------------------------------------------------------------------
// Acceptance #2 — Go verifies all four Rust-signed media types.
// ---------------------------------------------------------------------------

func TestProvenanceVerifyRustSignedEd25519Markdown(t *testing.T) {
	skipIfFixturesMissing(t)
	meta := loadProvenanceMetadata(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "rust_signed_ed25519.md")
	result, err := agent.VerifyText(target, &VerifyTextOpts{KeyDir: provenanceKeysDir(t)})
	if err != nil {
		t.Fatalf("VerifyText: %v", err)
	}
	if result.Status != "signed" {
		t.Fatalf("status=%q, want signed", result.Status)
	}
	if len(result.Signatures) != 1 {
		t.Fatalf("got %d sigs, want 1", len(result.Signatures))
	}
	sig := result.Signatures[0]
	if sig.Status != "valid" {
		t.Fatalf("sig status=%q, want valid", sig.Status)
	}
	if sig.Algorithm != "ed25519" {
		t.Fatalf("algorithm=%q, want ed25519", sig.Algorithm)
	}
	if sig.SignerID != meta.AgentEd25519.AgentID {
		t.Fatalf("signer_id=%q, want %q", sig.SignerID, meta.AgentEd25519.AgentID)
	}
}

func TestProvenanceVerifyRustSignedPq2025Markdown(t *testing.T) {
	skipIfFixturesMissing(t)
	meta := loadProvenanceMetadata(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "rust_signed_pq2025.md")
	result, err := agent.VerifyText(target, &VerifyTextOpts{KeyDir: provenanceKeysDir(t)})
	if err != nil {
		t.Fatalf("VerifyText: %v", err)
	}
	if result.Status != "signed" {
		t.Fatalf("status=%q, want signed", result.Status)
	}
	if len(result.Signatures) != 1 {
		t.Fatalf("got %d sigs, want 1", len(result.Signatures))
	}
	sig := result.Signatures[0]
	if sig.Status != "valid" {
		t.Fatalf("sig status=%q, want valid", sig.Status)
	}
	if sig.Algorithm != "pq2025" {
		t.Fatalf("algorithm=%q, want pq2025", sig.Algorithm)
	}
	if sig.SignerID != meta.AgentPq2025.AgentID {
		t.Fatalf("signer_id=%q, want %q", sig.SignerID, meta.AgentPq2025.AgentID)
	}
}

func TestProvenanceVerifyRustSignedMultiAlgo(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "rust_signed_multi_algo.md")
	result, err := agent.VerifyText(target, &VerifyTextOpts{KeyDir: provenanceKeysDir(t)})
	if err != nil {
		t.Fatalf("VerifyText: %v", err)
	}
	if result.Status != "signed" {
		t.Fatalf("status=%q, want signed", result.Status)
	}
	if len(result.Signatures) != 2 {
		t.Fatalf("got %d sigs, want 2", len(result.Signatures))
	}
	algos := []string{result.Signatures[0].Algorithm, result.Signatures[1].Algorithm}
	if (algos[0] != "ed25519" && algos[0] != "pq2025") ||
		(algos[1] != "ed25519" && algos[1] != "pq2025") ||
		algos[0] == algos[1] {
		t.Fatalf("expected one ed25519 and one pq2025; got %v", algos)
	}
	for _, sig := range result.Signatures {
		if sig.Status != "valid" {
			t.Fatalf("sig %v status=%q, want valid", sig.Algorithm, sig.Status)
		}
	}
}

func TestProvenanceVerifyRustSignedPng(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "rust_signed_ed25519.png")
	result, err := agent.VerifyImage(target, &VerifyImageOpts{KeyDir: provenanceKeysDir(t)})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if result.Status != "valid" {
		t.Fatalf("status=%q, want valid", result.Status)
	}
	if result.Format != "png" {
		t.Fatalf("format=%q, want png", result.Format)
	}
}

func TestProvenanceVerifyRustSignedJpeg(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "rust_signed_ed25519.jpg")
	result, err := agent.VerifyImage(target, &VerifyImageOpts{KeyDir: provenanceKeysDir(t)})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if result.Status != "valid" {
		t.Fatalf("status=%q, want valid", result.Status)
	}
	if result.Format != "jpeg" {
		t.Fatalf("format=%q, want jpeg", result.Format)
	}
}

func TestProvenanceVerifyRustSignedWebp(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "rust_signed_ed25519.webp")
	result, err := agent.VerifyImage(target, &VerifyImageOpts{KeyDir: provenanceKeysDir(t)})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if result.Status != "valid" {
		t.Fatalf("status=%q, want valid", result.Status)
	}
	if result.Format != "webp" {
		t.Fatalf("format=%q, want webp", result.Format)
	}
}

// ---------------------------------------------------------------------------
// C3 — yaml.v3 parses the Rust-signed markdown signature block body.
// ---------------------------------------------------------------------------

func TestProvenanceYamlParsesRustSignedBlockBody(t *testing.T) {
	skipIfFixturesMissing(t)

	content, err := os.ReadFile(filepath.Join(provenanceFixturesDir(t), "rust_signed_ed25519.md"))
	if err != nil {
		t.Fatal(err)
	}
	beginMarker := "-----BEGIN JACS SIGNATURE-----\n"
	endMarker := "\n-----END JACS SIGNATURE-----"
	beginIdx := strings.Index(string(content), beginMarker)
	if beginIdx < 0 {
		t.Fatal("no BEGIN marker found")
	}
	endIdx := strings.Index(string(content), endMarker)
	if endIdx < 0 {
		t.Fatal("no END marker found")
	}
	body := string(content[beginIdx+len(beginMarker) : endIdx])

	var parsed map[string]interface{}
	if err := yaml.Unmarshal([]byte(body), &parsed); err != nil {
		t.Fatalf("yaml.Unmarshal: %v", err)
	}
	for _, key := range []string{"signer", "signedContentHash", "publicKeyHash", "algorithm", "signature"} {
		if _, ok := parsed[key]; !ok {
			t.Fatalf("missing %q in parsed YAML body: %#v", key, parsed)
		}
	}
}

// ---------------------------------------------------------------------------
// C1 — strict + permissive parity for unsigned fixtures.
// ---------------------------------------------------------------------------

func TestProvenancePermissiveUnsignedMarkdown(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "unsigned.md")
	result, err := agent.VerifyText(target, nil)
	if err != nil {
		t.Fatalf("permissive verify: %v", err)
	}
	if result.Status != "missing_signature" {
		t.Fatalf("status=%q, want missing_signature", result.Status)
	}
}

func TestProvenanceStrictUnsignedMarkdown(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "unsigned.md")
	_, err = agent.VerifyText(target, &VerifyTextOpts{Strict: true})
	if err == nil {
		t.Fatal("expected strict verify to error")
	}
	if !errors.Is(err, ErrMissingSignature) {
		t.Fatalf("expected ErrMissingSignature, got %v", err)
	}
}

func TestProvenancePermissiveUnsignedImages(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	for _, fixture := range []struct {
		name string
		fmt  string
	}{
		{"unsigned.png", "png"},
		{"unsigned.jpg", "jpeg"},
		{"unsigned.webp", "webp"},
	} {
		t.Run(fixture.name, func(t *testing.T) {
			target := filepath.Join(provenanceFixturesDir(t), fixture.name)
			result, err := agent.VerifyImage(target, nil)
			if err != nil {
				t.Fatalf("permissive verify: %v", err)
			}
			if result.Status != "missing_signature" {
				t.Fatalf("status=%q, want missing_signature", result.Status)
			}
			if result.Format != fixture.fmt {
				t.Fatalf("format=%q, want %q", result.Format, fixture.fmt)
			}
		})
	}
}

func TestProvenanceStrictUnsignedImages(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	for _, fixture := range []string{"unsigned.png", "unsigned.jpg", "unsigned.webp"} {
		t.Run(fixture, func(t *testing.T) {
			target := filepath.Join(provenanceFixturesDir(t), fixture)
			_, err := agent.VerifyImage(target, &VerifyImageOpts{Strict: true})
			if err == nil {
				t.Fatal("expected strict verify to error")
			}
			if !errors.Is(err, ErrMissingSignature) {
				t.Fatalf("expected ErrMissingSignature, got %v", err)
			}
		})
	}
}

func TestProvenanceStrictRustSignedMarkdownDoesNotError(t *testing.T) {
	skipIfFixturesMissing(t)

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	target := filepath.Join(provenanceFixturesDir(t), "rust_signed_ed25519.md")
	result, err := agent.VerifyText(target, &VerifyTextOpts{
		Strict: true,
		KeyDir: provenanceKeysDir(t),
	})
	if err != nil {
		t.Fatalf("strict verify on signed file errored: %v", err)
	}
	if result.Status != "signed" {
		t.Fatalf("status=%q, want signed", result.Status)
	}
}

// ---------------------------------------------------------------------------
// Acceptance #2 — Go signs locally, Rust CLI verifies (round trip).
// ---------------------------------------------------------------------------

func TestProvenanceGoSignsRustVerifies(t *testing.T) {
	if _, err := exec.LookPath("cargo"); err != nil {
		t.Skip("cargo not on PATH; skipping Go→Rust round-trip")
	}

	tmp := t.TempDir()
	target := filepath.Join(tmp, "go_signed.md")
	if err := os.WriteFile(target, []byte("# Go-signed\n\nVerify me from Rust.\n"), 0o644); err != nil {
		t.Fatal(err)
	}

	agent, _, err := EphemeralSimpleAgent(strPtr("ed25519"))
	if err != nil {
		t.Fatal(err)
	}
	defer agent.Close()

	if err := agent.SignText(target, &SignTextOpts{NoBackup: true}); err != nil {
		t.Fatalf("SignText: %v", err)
	}

	signerID, err := agent.GetAgentID()
	if err != nil {
		t.Fatalf("GetAgentID: %v", err)
	}
	pem, err := agent.GetPublicKeyPEM()
	if err != nil {
		t.Fatalf("GetPublicKeyPEM: %v", err)
	}

	keyDir := filepath.Join(tmp, "keys")
	if err := os.Mkdir(keyDir, 0o755); err != nil {
		t.Fatal(err)
	}
	encoded := strings.ReplaceAll(strings.ReplaceAll(signerID, "..", "%2E%2E"), ":", "%3A")
	keyPath := filepath.Join(keyDir, encoded+".public.pem")
	if err := os.WriteFile(keyPath, []byte(pem), 0o644); err != nil {
		t.Fatal(err)
	}

	workspaceRoot := filepath.Clean(filepath.Join(provenanceFixturesDir(t), "..", "..", "..", ".."))

	cmd := exec.Command(
		"cargo", "run", "-q", "--bin", "jacs", "--",
		"verify-text", target,
		"--key-dir", keyDir,
		"--json",
	)
	cmd.Dir = workspaceRoot
	cmd.Env = append(os.Environ(), "JACS_MAX_IAT_SKEW_SECONDS=0")
	out, err := cmd.CombinedOutput()
	if err != nil {
		t.Fatalf("cargo verify-text failed: %v\noutput: %s", err, out)
	}

	// Filter out warning lines that cargo may emit before the JSON payload.
	jsonStart := strings.Index(string(out), "{")
	if jsonStart < 0 {
		t.Fatalf("no JSON payload in cargo output: %s", out)
	}
	var parsed struct {
		Status     string `json:"status"`
		Signatures []struct {
			Status   string `json:"status"`
			SignerID string `json:"signer_id"`
		} `json:"signatures"`
	}
	if err := json.Unmarshal(out[jsonStart:], &parsed); err != nil {
		t.Fatalf("unmarshal cli output: %v\noutput: %s", err, out)
	}
	if parsed.Status != "signed" {
		t.Fatalf("status=%q, want signed", parsed.Status)
	}
	if len(parsed.Signatures) != 1 {
		t.Fatalf("got %d signatures, want 1", len(parsed.Signatures))
	}
	sig := parsed.Signatures[0]
	if sig.Status != "valid" {
		t.Fatalf("sig status=%q, want valid", sig.Status)
	}
	if sig.SignerID != signerID {
		t.Fatalf("sig signer_id=%q, want %q", sig.SignerID, signerID)
	}
}

File: jacsnpm/test/cross-language-provenance.test.js
/**
 * Cross-language provenance tests (Task 13, PRD §5.1 / §5.2).
 *
 * Verifies that text + image fixtures signed by Rust under
 * `jacs/tests/fixtures/provenance/` are accepted by the Node binding,
 * and that a Node-signed file round-trips through the Rust `jacs verify-text`
 * CLI.
 *
 * Fixtures are committed; regenerate with:
 *
 *   UPDATE_PROVENANCE_FIXTURES=1 cargo test -p jacs --test \
 *     provenance_cross_language_tests -- --ignored regenerate_provenance_fixtures
 *
 * The test suite is skipped when fixtures are absent.
 */

const { expect } = require('chai');
const fs = require('fs');
const path = require('path');
const os = require('os');
const { spawnSync } = require('child_process');

let yaml;
try {
  yaml = require('js-yaml');
} catch (_) {
  yaml = null;
}

let JacsSimpleAgent;
try {
  ({ JacsSimpleAgent } = require('../index.js'));
} catch (_) {
  JacsSimpleAgent = null;
}

const FIXTURES_DIR = path.resolve(
  __dirname,
  '..',
  '..',
  'jacs',
  'tests',
  'fixtures',
  'provenance'
);
const KEYS_DIR = path.join(FIXTURES_DIR, 'keys');
const METADATA_PATH = path.join(FIXTURES_DIR, 'metadata.json');

function fixturesPresent() {
  return fs.existsSync(METADATA_PATH) && fs.existsSync(KEYS_DIR);
}

function readMetadata() {
  return JSON.parse(fs.readFileSync(METADATA_PATH, 'utf8'));
}

describe('cross-language provenance fixtures (Node verifies Rust-signed)', function () {
  this.timeout(60000);

  let agent;
  let metadata;

  before(async function () {
    if (!JacsSimpleAgent) {
      console.log('  Skipping: native binding not built');
      this.skip();
    }
    if (!fixturesPresent()) {
      console.log('  Skipping: provenance fixtures missing — run UPDATE_PROVENANCE_FIXTURES=1 cargo test ... --ignored regenerate_provenance_fixtures');
      this.skip();
    }
    metadata = readMetadata();
  });

  beforeEach(async function () {
    agent = await JacsSimpleAgent.ephemeral('ed25519');
  });

  // -------------------------------------------------------------------
  // Acceptance #2 — Node verifies all four Rust-signed media types.
  // -------------------------------------------------------------------

  it('verifies rust_signed_ed25519.md', async function () {
    const target = path.join(FIXTURES_DIR, 'rust_signed_ed25519.md');
    const result = await agent.verifyText(target, { keyDir: KEYS_DIR });
    expect(result.status).to.equal('signed');
    expect(result.signatures).to.have.lengthOf(1);
    const sig = result.signatures[0];
    expect(sig.status).to.equal('valid');
    expect(sig.algorithm).to.equal('ed25519');
    expect(sig.signer_id || sig.signerId).to.equal(metadata.agent_ed25519.agent_id);
  });

  it('verifies rust_signed_pq2025.md', async function () {
    const target = path.join(FIXTURES_DIR, 'rust_signed_pq2025.md');
    const result = await agent.verifyText(target, { keyDir: KEYS_DIR });
    expect(result.status).to.equal('signed');
    expect(result.signatures).to.have.lengthOf(1);
    const sig = result.signatures[0];
    expect(sig.status).to.equal('valid');
    expect(sig.algorithm).to.equal('pq2025');
    expect(sig.signer_id || sig.signerId).to.equal(metadata.agent_pq2025.agent_id);
  });

  it('verifies rust_signed_multi_algo.md (mixed-algorithm, unordered)', async function () {
    const target = path.join(FIXTURES_DIR, 'rust_signed_multi_algo.md');
    const result = await agent.verifyText(target, { keyDir: KEYS_DIR });
    expect(result.status).to.equal('signed');
    expect(result.signatures).to.have.lengthOf(2);

    const statuses = new Set(result.signatures.map((s) => s.status));
    expect([...statuses]).to.deep.equal(['valid']);

    const algos = result.signatures.map((s) => s.algorithm).sort();
    expect(algos).to.deep.equal(['ed25519', 'pq2025']);
  });

  it('verifies rust_signed_ed25519.png', async function () {
    const target = path.join(FIXTURES_DIR, 'rust_signed_ed25519.png');
    const result = await agent.verifyImage(target, { keyDir: KEYS_DIR });
    expect(result.status).to.equal('valid');
    expect(result.format).to.equal('png');
  });

  it('verifies rust_signed_ed25519.jpg', async function () {
    const target = path.join(FIXTURES_DIR, 'rust_signed_ed25519.jpg');
    const result = await agent.verifyImage(target, { keyDir: KEYS_DIR });
    expect(result.status).to.equal('valid');
    expect(result.format).to.equal('jpeg');
  });

  it('verifies rust_signed_ed25519.webp', async function () {
    const target = path.join(FIXTURES_DIR, 'rust_signed_ed25519.webp');
    const result = await agent.verifyImage(target, { keyDir: KEYS_DIR });
    expect(result.status).to.equal('valid');
    expect(result.format).to.equal('webp');
  });

  // -------------------------------------------------------------------
  // C3 — js-yaml parses the Rust-signed markdown signature block body
  // as a full JACS YAML document footer.
  // -------------------------------------------------------------------

  it('C3: js-yaml parses Rust-signed markdown block body as full JACS footer', function () {
    if (!yaml) this.skip();
    const content = fs.readFileSync(
      path.join(FIXTURES_DIR, 'rust_signed_ed25519.md'),
      'utf8'
    );
    const begin = '-----BEGIN JACS SIGNATURE-----\n';
    const endMarker = '\n-----END JACS SIGNATURE-----';
    const start = content.indexOf(begin) + begin.length;
    const end = content.indexOf(endMarker);
    expect(end).to.be.greaterThan(start);
    const body = content.slice(start, end);
    const parsed = yaml.load(body);

    expect(parsed).to.have.property('jacsType', 'inline-md');
    expect(parsed).to.have.property('jacsId').that.is.a('string');
    expect(parsed).to.have.property('jacsVersion').that.is.a('string');
    expect(parsed).to.have.nested.property('jacsSignature.agentID').that.is.a('string');
    expect(parsed).to.have.nested.property('jacsSignature.publicKeyHash').that.is.a('string');
    expect(parsed).to.have.nested.property('jacsSignature.signature').that.is.a('string');
    expect(parsed).to.have.nested.property('content.inlineSignatureVersion', 1);
    expect(parsed).to.have.nested.property('content.signedContentHash').that.is.a('string');
  });

  // -------------------------------------------------------------------
  // C1 — strict + permissive parity for unsigned fixtures.
  // -------------------------------------------------------------------

  it('C1 permissive: unsigned.md returns missing_signature', async function () {
    const target = path.join(FIXTURES_DIR, 'unsigned.md');
    const result = await agent.verifyText(target);
    expect(result.status).to.equal('missing_signature');
    // The binding omits `signatures` for the missing case (matches the wire
    // shape produced by binding-core::serialize_verify_text_result). Either
    // unset or [] is acceptable for downstream consumers.
    expect(result.signatures || []).to.deep.equal([]);
  });

  it('C1 strict: unsigned.md rejects /no JACS signature found/', async function () {
    const target = path.join(FIXTURES_DIR, 'unsigned.md');
    let caught;
    try {
      await agent.verifyText(target, { strict: true });
    } catch (e) {
      caught = e;
    }
    expect(caught, 'expected strict verifyText to reject').to.exist;
    expect(caught.message).to.match(/no JACS signature found/);
  });

  for (const [fixture, expectedFmt] of [
    ['unsigned.png', 'png'],
    ['unsigned.jpg', 'jpeg'],
    ['unsigned.webp', 'webp'],
  ]) {
    it(`C1 permissive: ${fixture} returns missing_signature`, async function () {
      const target = path.join(FIXTURES_DIR, fixture);
      const result = await agent.verifyImage(target);
      expect(result.status).to.equal('missing_signature');
      // format detection still happens before the signature check.
      expect(result.format).to.equal(expectedFmt);
    });

    it(`C1 strict: ${fixture} rejects /no JACS signature found/`, async function () {
      const target = path.join(FIXTURES_DIR, fixture);
      let caught;
      try {
        await agent.verifyImage(target, { strict: true });
      } catch (e) {
        caught = e;
      }
      expect(caught, `expected strict verifyImage to reject ${fixture}`).to.exist;
      expect(caught.message).to.match(/no JACS signature found/);
    });
  }

  it('strict on rust_signed_ed25519.md still resolves with status=signed', async function () {
    const target = path.join(FIXTURES_DIR, 'rust_signed_ed25519.md');
    const result = await agent.verifyText(target, {
      strict: true,
      keyDir: KEYS_DIR,
    });
    expect(result.status).to.equal('signed');
  });

  // -------------------------------------------------------------------
  // Acceptance #2 — Node signs locally, Rust CLI verifies (round trip).
  // -------------------------------------------------------------------

  it('node-signs then Rust CLI verifies (round trip)', async function () {
    const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'jacs-xlang-'));
    try {
      const target = path.join(tmp, 'node_signed.md');
      fs.writeFileSync(target, '# Node-signed\n\nVerify me from Rust.\n');

      await agent.signText(target);

      const keyDir = path.join(tmp, 'keys');
      fs.mkdirSync(keyDir);
      const signerId = await agent.getAgentId();
      const pem = await agent.getPublicKeyPem();
      const encoded = signerId.replace(/\.\./g, '%2E%2E').replace(/:/g, '%3A');
      fs.writeFileSync(path.join(keyDir, `${encoded}.public.pem`), pem);

      const workspaceRoot = path.resolve(__dirname, '..', '..');
      // Prefer cargo run from a developer checkout (matches the Python test
      // path); the CI image always has cargo available.
      const result = spawnSync(
        'cargo',
        [
          'run',
          '-q',
          '--bin',
          'jacs',
          '--',
          'verify-text',
          target,
          '--key-dir',
          keyDir,
          '--json',
        ],
        {
          cwd: workspaceRoot,
          encoding: 'utf8',
          env: { ...process.env, JACS_MAX_IAT_SKEW_SECONDS: '0' },
        }
      );
      expect(
        result.status,
        `cargo verify-text exited ${result.status}\nstdout: ${result.stdout}\nstderr: ${result.stderr}`
      ).to.equal(0);

      const parsed = JSON.parse(result.stdout);
      expect(parsed.status).to.equal('signed');
      expect(parsed.signatures).to.have.lengthOf(1);
      expect(parsed.signatures[0].status).to.equal('valid');
      expect(parsed.signatures[0].signer_id).to.equal(signerId);
    } finally {
      fs.rmSync(tmp, { recursive: true, force: true });
    }
  });
});
